packages/core/src/browser/label-provider.ts

```diff
@@ -154,6 +196,122 @@ export class DefaultUriLabelProviderContribution implements LabelProviderContrib
     protected getUri(element: URI | URIIconReference): URI | undefined {
         return URIIconReference.is(element) ? element.uri : element;
     }
+
+    registerFormatter(formatter: ResourceLabelFormatter): Disposable {
+        this.formatters.push(formatter);
+        this.onDidChangeEmitter.fire({
+            affects: (element: URI) => this.canHandle(element) === 1
+        });
+        return Disposable.create(() => {
+            this.formatters = this.formatters.filter(f => f !== formatter);
+            this.onDidChangeEmitter.fire({
+                affects: (element: URI) => this.canHandle(element) === 1
+            });
+        });
+    }
+
+    getUriLabel?(uri: URI): string {
+        const formatting = this.findFormatting(uri);
+        if (!formatting) {
+            return uri.path.toString();
+        }
+        return this.formatUri(uri, formatting);
+    }
+
+    get onDidChange(): Event<DidChangeLabelEvent> {
+        return this.onDidChangeEmitter.event;
+    }
+
+    // copied and modified from https://github.com/microsoft/vscode/blob/1.44.2/src/vs/workbench/services/label/common/labelService.ts#L240-L274
+    /*---------------------------------------------------------------------------------------------
+    *  Copyright (c) Microsoft Corporation. All rights reserved.
+    *  Licensed under the MIT License. See License.txt in the project root for license information.
+    *--------------------------------------------------------------------------------------------*/
+    private readonly labelMatchingRegexp = /\${(scheme|authority|path|query)}/g;
+    private formatUri(resource: URI, formatting: ResourceLabelFormatting): string {
+        // const labelMatchingRegexp = /\${(scheme|authority|path|query)}/g;
+        let label = formatting.label.replace(this.labelMatchingRegexp, (match, token) => {
+            switch (token) {
+                case 'scheme': return resource.scheme;
+                case 'authority': return resource.authority;
+                case 'path': return resource.path.toString();
+                case 'query': return resource.query;
+                default: return '';
+            }
+        });
+
+        // convert \c:\something => C:\something
+        if (formatting.normalizeDriveLetter && this.hasDriveLetter(label)) {
+            label = label.charAt(1).toUpperCase() + label.substr(2);
+        }
+
+        if (formatting.tildify) {
+            const homePath = process.env.HOMEPATH;
```

Comment:
it won't work, we are in browser, one should get home from `EnvVarServer` after https://github.com/eclipse-theia/theia/pull/7908 is merged

Reply:
Used `EnvVariablesServer` in the constructor.